ThomazFB/WordPress-Android
# Tenor API integrated with GIF Picker

Summary
------------
This Pull Request brings integration for the Tenor API within the pre-existent GIF selection code inside the app. As described at [issue 11456](https://github.com/wordpress-mobile/WordPress-Android/issues/11456).

![screencapture-1585095917110 2020-03-24 23_04_23](https://user-images.githubusercontent.com/5920403/77494177-d3876d00-6e23-11ea-9928-16cf37e264e3.gif)

The solution proposes the following implementation:

* A `GifProvider` interface must be the only coupling between all code for the GIF picker and the Tenor implementation, avoiding any reference to the Tenor library and making it easy to replace the Provider as we're doing now

* The `TenorProvider` implements the `GifProvider` interface and is injected within the `DataSource` via Dagger, this way only the `AppModule` knows about the `TenorProvider`. Also, the `TenorProvider` is covered by unit tests in the `TenorProviderTest` class.

* The already implemented pagination feature and list actions are all kept and integrated to work smoothly with the `GifProvider`, this integration happens through the available `PositionalDataSource`, The Tenor implementation makes sure that positioning tracking is respected

* The original integration with the `MediaBrowserActivity` was restored, making GIF selection available again

Known missing points for upcoming Pull Requests
------------
* Integration with `EditPostActivity` not restored yet
* All Giphy packages and class naming is still missing, this will be the last modification to not storm the code review with name changes within the Pull Request
* Change Giphy icons
* Analytics support still to be added
* Unit tests for PickerDataSource

Testing
------------
1. Create a Tenor API key and paste it associated with the `wp.tenor.api_key` as the value inside your `gradle.properties`
2. Go to the Media Browser
3. Click on the Add button at the top right. Select the `Choose from Tenor` option
4. Type in the search bar to query GIFs
5. Once inside the Gif Picker, make sure that `selection`, `preview` and `add` are all working as expected


PR submission checklist
------------
